### PR TITLE
[FSDP2] Accumulated in `reduce_dtype` if not syncing grads

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #125191
* #125190

For microbatching use cases (e.g. PP), we may use fp32 reduce-scatter (i.e. `MixedPrecisionPolicy(reduce_dtype=torch.float32)`), where we want to accumulate the unsharded gradients in fp32 across microbatches until reduce-scattering in fp32 upon the last microbatch.

Note that the `unsharded_param` is in bf16, so we must save the fp32 accumulated gradient to an attribute different from `.grad`. Moreover, saving a new attribute on the `torch.Tensor` leads to some annoying type checking issues (where the attribute may not be defined), so this PR prefers to save the attribute on the `FSDPParam` class instead.

One could argue that this behavior should be configurable, but since I think for large-scale training, everyone is leaning toward fp32 accumulation across microbatches, let us avoid adding another argument for now.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

### CODE REVIEW

**Review by weifengpy — APPROVED** [review]
I had a confusion about cpu offloading + `fsdp_param.unsharded_accumulated_grad_data` but it turns out does not matter. cpu offloading is only applied to shareded params

File: torch/distributed/_composable/fsdp/_fsdp_param.py [hunk 1]
```diff
@@ -126,6 +126,7 @@ class FSDPParam:
     _sharded_post_forward_param_data: Optional[torch.Tensor]  # 1D
     _sharded_post_forward_param: Optional[nn.Parameter]  # ND
     _unsharded_param: nn.Parameter  # ND
+    unsharded_accumulated_grad: Optional[torch.Tensor]  # ND
```
Comment (wanchaol) [inline torch/distributed/_composable/fsdp/_fsdp_param.py:129]: nit: keep this be a private variable to align with all other fields?

Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_param.py:129]: Sorry, I think the notion of private and public is non-obvious for the `FSDPParam` class. The way I have is is that if `FSDPParamGroup` should access an attribute on `FSDPParam`, then that attribute should be public. (Note that `FSDPParam` will never be accessed publicly by the user, as there is no way to do that today and should not be a way in the future.)

For this attribute and the current implementation, `FSDPParamGroup` needs to check it if it is `None` or not to know if there is an unsharded accumulated gradient to reduce-scatter or not, so I have it as public.

**Comment by awgu** [comment]
Failures were all from the open registration test.

**Pull request closed without merging**